File: library-core/src/main/java/com/mikepenz/iconics/animation/BlinkScaleProcessor.java
/*
 * Copyright 2018 Mike Penz
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mikepenz.iconics.animation;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.text.TextPaint;

import androidx.annotation.FloatRange;
import androidx.annotation.NonNull;

import com.mikepenz.iconics.IconicsBrush;

/**
 * @author pa.gulko zTrap (30.11.2018)
 */
public class BlinkScaleProcessor extends IconicsAnimationProcessor {
    /**
     * Duration used for all instances of this processor by default. 500 ms by default.
     */
    public static int defaultDuration = 500;

    private @FloatRange(from = 0)
    float mMinimumScale = 0;
    private @FloatRange(from = 0)
    float mMaximumScale = 1;

    {
        mRepeatMode = REVERSE;
        mDuration = defaultDuration;
    }

    /**
     * @param minimumScale The scale which will be used as minimal available value.
     */
    public @NonNull
    BlinkScaleProcessor minimumScale(@FloatRange(from = 0) float minimumScale) {
        mMinimumScale = minimumScale;
        return this;
    }

    /**
     * @param maximumScale The scale which will be used as maximal available value.
     */
    public @NonNull
    BlinkScaleProcessor maximumScale(@FloatRange(from = 0) float maximumScale) {
        mMaximumScale = maximumScale;
        return this;
    }

    /**
     * @return The minimal available scale.
     */
    public @FloatRange(from = 0)
    float getMinimumScale() {
        return mMinimumScale;
    }

    /**
     * @return The minimal available scale.
     */
    public @FloatRange(from = 0)
    float getMaximumScale() {
        return mMaximumScale;
    }

    @Override
    public @NonNull
    String animationTag() {
        return "blink_scale";
    }

    @Override
    protected void processPreDraw(
            @NonNull Canvas canvas,
            @NonNull IconicsBrush<TextPaint> iconBrush,
            @NonNull IconicsBrush<Paint> iconContourBrush,
            @NonNull IconicsBrush<Paint> backgroundBrush,
            @NonNull IconicsBrush<Paint> backgroundContourBrush) {

        float scaleByPercent = (mMaximumScale - mMinimumScale) / 100;

        float scale = getAnimatedPercent() * scaleByPercent;
        Rect bounds = getDrawableBounds();

        canvas.save();
        canvas.scale(scale, scale, bounds.width() / 2, bounds.height() / 2);
    }

    @Override
    protected void processPostDraw(@NonNull Canvas canvas) {
        canvas.restore();
    }
}

File: library-core/src/main/java/com/mikepenz/iconics/animation/BlinkScaleProcessor.kt
/*
 * Copyright 2019 Mike Penz
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mikepenz.iconics.animation

import android.animation.TimeInterpolator
import android.graphics.Canvas
import android.graphics.Paint
import android.text.TextPaint
import androidx.annotation.FloatRange
import com.mikepenz.iconics.IconicsBrush
import com.mikepenz.iconics.animation.IconicsAnimationProcessor.RepeatMode.REVERSE

/**
 * @author pa.gulko zTrap (30.11.2018)
 */
open class BlinkScaleProcessor(
    /** The minimal available scale. */
    @FloatRange(from = 0.0) open var minimumScale: Float = 0f,
    /** The maximal available scale. */
    @FloatRange(from = 0.0) open var maximumScale: Float = 1f,

    override var interpolator: TimeInterpolator = DEFAULT_INTERPOLATOR,
    override var duration: Long = DEFAULT_DURATION,
    override var repeatCount: Int = INFINITE,
    override var repeatMode: IconicsAnimationProcessor.RepeatMode = REVERSE,
    override var isStartImmediately: Boolean = true
) : IconicsAnimationProcessor(interpolator, duration, repeatCount, repeatMode, isStartImmediately) {

    companion object {
        /** Duration used for all instances of this processor by default. 500 ms by default. */
        @JvmField var DEFAULT_DURATION: Long = 500L
    }

    override val animationTag: String = "blink_scale"

    override fun processPreDraw(
        canvas: Canvas,
        iconBrush: IconicsBrush<TextPaint>,
        iconContourBrush: IconicsBrush<Paint>,
        backgroundBrush: IconicsBrush<Paint>,
        backgroundContourBrush: IconicsBrush<Paint>
    ) {

        val scaleByPercent = (maximumScale - minimumScale) / 100

        val scale = animatedPercent * scaleByPercent
        val bounds = drawableBounds

        bounds?.let {
            canvas.save()
            canvas.scale(scale, scale, (it.width() / 2).toFloat(), (it.height() / 2).toFloat())
        }
    }

    override fun processPostDraw(canvas: Canvas) {
        canvas.restore()
    }
}
